# IGVZ Intervention Study - Effect on Migrant Women

In [1]:
import pandas as pd
from scipy.stats import mannwhitneyu, chi2_contingency

# Load the dataset
data_path = '/Users/nagesh/Library/CloudStorage/OneDrive-NextGenMetricsB.V/Personal/Anisha/M3/data.xlsx'
try:
    # Attempt to read the Excel file
    data = pd.read_excel(data_path)
    # If successful, display the first few rows of the dataframe
    message = "Excel file read successfully! Here are the first few rows of the data:"
    preview = data.head()
except Exception as e:
    # If reading fails, display an error message
    message = f"An error occurred: {e}"
    preview = None

data


groupID  Patient_ID  IGVZ_intervention  Years_of_education  \
0          1     2018001                  1                  99   
1          1     2018002                  1                  99   
2          1     2018003                  1                  99   
3          1     2018004                  1                  12   
4          1     2018005                  1                  12   
..       ...         ...                ...                 ...   
125        0     2020081                  0                  12   
126        0     2020083                  0                   0   
127        0     2020087                  0                  10   
128        0     2020090                  0                  18   
129        0     2020092                  0                   8   

     Country_of_origin  Gravida  Para  T1_Gestational_age  T1_KQ_1  T1_KQ_2  \
0                    1        2     1                37.0        2        1   
1                    2        1     0                28.0        2        3   
2                    3        1     0                99.0        2        3   
3                    4        1     0                38.0        3        1   
4                    3        1     0                35.0        1        3   
..                 ...      ...   ...                 ...      ...      ...   
125                  3        1     0                27.0        2        3   
126                  3        5    99                30.0        3        1   
127                  3        2    99                32.0        1        3   
128                  3        5    99                20.0        2        1   
129                 13        1     0                33.0        3        3   

     ...  T2_KQ_13a_Score  T2_KQ_13b_Score  T2_KQ_13c_Score  T2_KQ_13d_Score  \
0    ...                0                0                0                0   
1    ...                1                1                0                1   
2    ...                0                0                0                0   
3    ...                0                0                0                0   
4    ...                0                0                0                0   
..   ...              ...              ...              ...              ...   
125  ...                1                0                0                0   
126  ...                1                1                1                0   
127  ...                1                1                1                1   
128  ...                0                0                0                0   
129  ...                0                1                0                0   

     T2_KQ_13e_Score  T2_KQ_13f_Score  T2_KQ_13g_Score  T2_KQ_13h_Score  \
0                  0                0                0                0   
1                  0                1                1                0   
2                  0                0                0                0   
3                  0                0                0                0   
4                  0                0                0                0   
..               ...              ...              ...              ...   
125                0                0                0                0   
126                0                0                1                1   
127                1                1                1                1   
128                0                0                0                0   
129                0                1                0                1   

     T2_KQ_13i_Score  T2_KQ_Total_Score  
0                  0                  0  
1                  1                 14  
2                  0                  0  
3                  0                  0  
4                  0                  0  
..               ...                ...  
125                0                  5  
126                1                 12  
127        

In [2]:
# Display info with verbose=False
data.info(verbose=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 115 columns):
 #    Column              Dtype  
---   ------              -----  
 0    groupID             int64  
 1    Patient_ID          int64  
 2    IGVZ_intervention   int64  
 3    Years_of_education  int64  
 4    Country_of_origin   int64  
 5    Gravida             int64  
 6    Para                int64  
 7    T1_Gestational_age  float64
 8    T1_KQ_1             int64  
 9    T1_KQ_2             int64  
 10   T1_KQ_3             int64  
 11   T1_KQ_4             int64  
 12   T1_KQ_5             int64  
 13   T1_KQ_6             int64  
 14   T1_KQ_7             int64  
 15   T1_KQ_8             int64  
 16   T1_KQ_9             int64  
 17   T1_KQ_10            int64  
 18   T1_KQ_11            int64  
 19   T1_KQ_12            int64  
 20   T1_KQ_13a           int64  
 21   T1_KQ_13b           int64  
 22   T1_KQ_13c           int64  
 23   T1_KQ_13d           int64  
 24   T1_K

Confirm the number of Treatment Group v/s Control Group participants 

In [ ]:
# Confirm the number of members in the treatment group (groupID == 1)
treatment_group_count = data[data['groupID'] == 1].shape[0]
print('treatment group participants:', treatment_group_count)

control_group_count = data[data['groupID'] == 0].shape[0]
print('control group participants:', control_group_count)


In [ ]:
import pandas as pd
from scipy.stats import mannwhitneyu, chi2_contingency

# Load your data
#data = pd.read_csv('path_to_your_csv_file.csv')

# Define the parameters to compare
parameters = ['Years_of_education', 'Country_of_origin', 'Gravida', 'Para', 'T1_Gestational_age']

# Assuming 'groupID' column with 1 for treatment and 0 for control group
treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Initialize a dictionary to store p-values
p_values = {}

# Iterate over the parameters and perform the appropriate statistical test
for param in parameters:
    if param == 'Country_of_origin':  # Assuming 'Country_of_origin' should be treated as categorical
        # Create a contingency table
        contingency_table = pd.crosstab(data['groupID'], data[param])
        chi2, p, dof, expected = chi2_contingency(contingency_table)
        p_values[param] = p
    else:
        # Assume other variables are continuous and use Mann-Whitney U test
        stat, p = mannwhitneyu(treatment_group[param], control_group[param], alternative='two-sided')
        p_values[param] = p

# Output the p-values
print(p_values)


# Shapiro-Wilk test for detecting Normality in distribution 

The Excel sheet has been read successfully. It contains data including group IDs, patient IDs, information on whether the IGVZ intervention was received, years of education, country of origin, gravida, para, gestational age at two time points (T1 and T2), and scores on a knowledge questionnaire at both time points, among other variables.

We can now proceed with the Shapiro-Wilk test for normality on the appropriate variables.

In [ ]:
from scipy.stats import shapiro

# Identify columns related to T1 scores
t1_score_columns = [col for col in data.columns if 'T1_' in col and '_Score' in col]

# Initialize a dictionary to hold the p-values of the Shapiro-Wilk test for normality
shapiro_p_values = {}

# Perform Shapiro-Wilk test for each T1 score column and store the p-values
for col in t1_score_columns:
    shapiro_test = shapiro(data[col])
    shapiro_p_values[col] = "{:.17f}".format(shapiro_test.pvalue)

shapiro_p_values


T2 Scores (Post Intervention)

In [ ]:
from scipy.stats import shapiro

# Identify columns related to T1 scores
t1_score_columns = [col for col in data.columns if 'T2_' in col and '_Score' in col]

# Initialize a dictionary to hold the p-values of the Shapiro-Wilk test for normality
shapiro_p_values = {}

# Perform Shapiro-Wilk test for each T1 score column and store the p-values
for col in t1_score_columns:
    shapiro_test = shapiro(data[col])
    shapiro_p_values[col] = "{:.17f}".format(shapiro_test.pvalue)

shapiro_p_values


If the Shapiro-Wilk test on the T2 scores also yields non-significant results (i.e., p-values less than your chosen significance level, often 0.05), it means that the data for the T2 knowledge scores also deviates from a normal distribution.

In simpler terms, imagine you're sorting candies into different colored jars. For normally distributed data, the candies would be fairly evenly spread out across the jars. But if the data is not normally distributed, it's like most of the candies end up in one or two jars, and very few candies are in the other jars. This uneven distribution is what the Shapiro-Wilk test detects.

When the data is not normally distributed, it suggests that the scores at T2 for both treatment and control groups don't follow the typical bell-shaped curve you might expect. Instead, the scores might be clustered at certain levels, with fewer participants scoring at the extremes.

For analysis, this means you might want to use non-parametric tests, like the Mann-Whitney U test, to compare the T2 knowledge scores between the treatment and control groups. These tests don't rely on the assumption of normality and are robust to non-normal distributions. They assess differences in the rank ordering of scores rather than their distributional shape.

In essence, if the T2 scores are also not normally distributed, you can still analyze the data using appropriate statistical methods. The key is to choose tests that don't require the assumption of normality and are suitable for the type of data you have.

The Shapiro-Wilk test results for the T2 scores indicate extremely small p-values for each of the variables, with values ranging from 0.00000000000000000 to 0.00000008817142739. This indicates that the null hypothesis of normality is rejected for all T2 score columns.

In simpler terms, it means that the data for the T2 knowledge scores does not follow a normal distribution. This suggests that the scores at T2 for both treatment and control groups are not spread out evenly but are instead clustered at certain levels, with fewer participants scoring at the extremes.

Given that the T2 scores are not normally distributed, you would want to use non-parametric tests, such as the Mann-Whitney U test, to compare the T2 knowledge scores between the treatment and control groups. These tests do not rely on the assumption of normality and are suitable for analyzing non-normally distributed data.

When interpreting the results of the Mann-Whitney U test, focus on comparing the median scores between the treatment and control groups rather than the means, as the test assesses differences in the rank ordering of scores rather than their distributional shape.


Shapiro-Wilk Test for residuals

In [ ]:
# Step 1: Calculate the change in total knowledge scores from T1 to T2 for each participant
data['Change_in_KQ_Total_Score'] = data['T2_KQ_Total_Score'] - data['T1_KQ_Total_Score']

# Step 2: Calculate the mean change for both the intervention group and control group
mean_change_intervention = data[data['groupID'] == 1]['Change_in_KQ_Total_Score'].mean()
mean_change_control = data[data['groupID'] == 0]['Change_in_KQ_Total_Score'].mean()

# Step 3: Compute residuals
data['Residuals'] = data.apply(lambda row: row['Change_in_KQ_Total_Score'] - (mean_change_intervention if row['groupID'] == 1 else mean_change_control), axis=1)

# Step 4: Perform the Shapiro-Wilk test on these residuals
shapiro_test_resid = shapiro(data['Residuals'])
shapiro_test_resid


The Shapiro-Wilk test result shows two key pieces of information:

Statistic = 0.4353: This is the test statistic value, which measures how closely your data's distribution matches a normal distribution. Values closer to 1 suggest a distribution closer to normal.
P-value = 3.07e-20: This is an extremely small number, practically zero, which indicates the probability of observing your data if it were drawn from a normal distribution.
Given that the p-value is much less than the typical alpha level of 0.05, you can reject the null hypothesis of the Shapiro-Wilk test, which states that the data is normally distributed. This means that the residuals from the changes in total knowledge scores between T1 and T2 for your groups do not follow a normal distribution.

How to Proceed Further

Given the non-normal distribution of your residuals, proceeding with parametric tests that assume normality (like the two-way repeated measures ANOVA) might not be the most appropriate choice. However, if two-way repeated measures ANOVA is "robust" to violations of normality, meaning it can still provide valid results when the data is not perfectly normal but close (which is clearly not the case)

the situation, indicated by the Shapiro-Wilk test result, shows a significant deviation from normality. Here are some suggestions on how to proceed:

Transformation of Data: Consider transforming your data to achieve normality. Common transformations include log, square root, or inverse. After transformation, you can perform the Shapiro-Wilk test again to check for normality.

Non-Parametric Alternatives: Since your data significantly deviates from normality, look for non-parametric alternatives. For repeated measures data, the Friedman test is a non-parametric alternative for comparing more than two related groups. However, for analyzing the interaction effect (group x time) without assuming normality, options are more limited.

Robust Methods: Explore robust statistical methods that are less sensitive to violations of normality assumptions. These methods provide ways to analyze data that do not meet the assumptions required for traditional parametric tests.

Consult with Your Supervisor: Share these findings and discuss the significant deviation from normality. Your supervisor may agree to proceed with the analysis acknowledging the data's limitations or may suggest an alternative approach.

Report Findings: Regardless of the chosen method, clearly report the findings from the Shapiro-Wilk test in your methods section. Explain how the results influenced your choice of statistical tests and discuss the implications for interpreting your study's findings.

Considering the advice to use a two-way repeated measures ANOVA, you might still discuss its use with your supervisor, especially focusing on its robustness and any potential impact the violation of normality could have on your results.

# Transform the total scores using Square Root transformation

In [ ]:
import numpy as np

# Apply a square root transformation to the total knowledge scores at T1 and T2
sqrt_transformed_T1_scores = np.sqrt(data['T1_KQ_Total_Score'])
sqrt_transformed_T2_scores = np.sqrt(data['T2_KQ_Total_Score'])

# Perform the Shapiro-Wilk test on the transformed data for both T1 and T2
shapiro_test_T1_sqrt_transformed = shapiro(sqrt_transformed_T1_scores)
shapiro_test_T2_sqrt_transformed = shapiro(sqrt_transformed_T2_scores)

(shapiro_test_T1_sqrt_transformed, shapiro_test_T2_sqrt_transformed)


The results of the Shapiro-Wilk test on your square root-transformed total knowledge scores for T1 and T2 indicate that both transformed distributions still deviate significantly from normality:

T1 Transformed Scores:

Statistic: 0.6782
P-value: 1.68e-15
T2 Transformed Scores:

Statistic: 0.7451
P-value: 9.77e-14
In both cases, the test statistics (0.6782 for T1 and 0.7451 for T2) are far from 1, indicating a departure from normality. Furthermore, the extremely small p-values (far below the conventional alpha level of 0.05) lead us to reject the null hypothesis that the data is normally distributed.

Implications and Next Steps

Given these results, it's clear that even after a square root transformation, the distributions of your total knowledge scores at T1 and T2 are not normal. This suggests a need for either further transformation attempts or the consideration of non-parametric statistical tests that do not assume normality.

Further Transformation Attempts: You could try other transformations, such as the logarithmic (if applicable), Box-Cox (if all values are positive), or even more robust methods like rank transformations. Each has its own conditions and suitability depending on the nature of the data's skewness and scale.

Non-Parametric Tests: Given the challenges in achieving a normal distribution through transformations, non-parametric alternatives to your intended analyses might be more appropriate. These tests do not assume normality. For comparing two related samples (e.g., T1 vs. T2 scores within the same subjects), the Wilcoxon signed-rank test is a common choice. For comparing two independent samples (e.g., intervention vs. control groups), the Mann-Whitney U test is widely used.


# Transform the total scores using Logarithmic transform

In [ ]:
# Check for non-positive values in T1 and T2 total knowledge scores
non_positive_T1 = any(data['T1_KQ_Total_Score'] <= 0)
non_positive_T2 = any(data['T2_KQ_Total_Score'] <= 0)

# Choose a transformation based on the presence of non-positive values
if non_positive_T1 or non_positive_T2:
    # If there are non-positive values, we will use a square root transformation after adding a constant to ensure positivity
    constant = 1 - min(data['T1_KQ_Total_Score'].min(), data['T2_KQ_Total_Score'].min())
    transformed_T1_scores = np.sqrt(data['T1_KQ_Total_Score'] + constant)
    transformed_T2_scores = np.sqrt(data['T2_KQ_Total_Score'] + constant)
else:
    # If all values are positive, we can directly apply a logarithmic transformation
    transformed_T1_scores = np.log(data['T1_KQ_Total_Score'])
    transformed_T2_scores = np.log(data['T2_KQ_Total_Score'])

# Perform the Shapiro-Wilk test on the transformed data for both T1 and T2
shapiro_test_T1_transformed = shapiro(transformed_T1_scores)
shapiro_test_T2_transformed = shapiro(transformed_T2_scores)

(shapiro_test_T1_transformed, shapiro_test_T2_transformed)


The Shapiro-Wilk test results for your logarithmically transformed total knowledge scores at T1 and T2 still indicate significant deviations from a normal distribution:

T1 Log-Transformed Scores:

Statistic: 0.6499
P-value: 3.62e-16
T2 Log-Transformed Scores:

Statistic: 0.8020
P-value: 5.83e-12
Even after the logarithmic transformation, the test statistics (0.6499 for T1 and 0.8020 for T2) are below the threshold for normality, and the p-values are well below the commonly used alpha level of 0.05. This leads to the rejection of the null hypothesis that the data comes from a normally distributed population.

Interpretation and Recommendations:
Given that both square root and logarithmic transformations have not resulted in distributions that satisfy the criteria for normality, it suggests that your data inherently has properties that resist these transformations to normality. This could be due to extreme values, skewness, or the nature of the data itself.

Non-Parametric Tests: At this point, considering non-parametric alternatives for your analysis would be advisable. These tests do not assume normal distribution and can provide robust insights even when parametric assumptions are not met. For comparisons within the same group over time, consider the Wilcoxon signed-rank test. For comparing independent groups, the Mann-Whitney U test could be suitable.

Further Data Exploration: Beyond transformation and testing for normality, exploring your data for outliers, data entry errors, or subgroups that may behave differently could offer insights. Understanding the nature of your data more deeply can inform both the choice of statistical tests and the interpretation of your results.

Data Transformation and Analysis Reporting: In your reporting, it's essential to detail the steps taken to assess normality and the rationale behind the chosen statistical methods. Transparency about the challenges encountered and how they were addressed adds to the credibility and reproducibility of your research.


# Lets say we choose Parametric tests - JUST HYPOTHETICALLY

In [ ]:
import statsmodels.api as sm
from statsmodels.stats.anova import AnovaRM

# Prepare the data for two-way repeated measures ANOVA
anova_data = data[['Patient_ID', 'groupID', 'T1_KQ_Total_Score', 'T2_KQ_Total_Score']].melt(id_vars=['Patient_ID', 'groupID'],
                                                                                              var_name='Time',
                                                                                              value_name='Total_Score')
anova_data['Time'] = anova_data['Time'].map({'T1_KQ_Total_Score': 'T1', 'T2_KQ_Total_Score': 'T2'})

# Encode 'groupID' as a categorical variable for clarity
anova_data['groupID'] = anova_data['groupID'].astype('category')

# Assuming 'anova_data' is your DataFrame prepared for ANOVA
# Aggregate the data to ensure one observation per subject per condition
#aggregated_data = anova_data.groupby(['Patient_ID', 'Time', 'groupID']).mean().fillna(0).reset_index()
aggregated_data = anova_data.groupby(['Patient_ID', 'Time', 'groupID'], observed=False).mean().fillna(0).reset_index()

# Perform two-way repeated measures ANOVA
try:
# Then, attempt the two-way repeated measures ANOVA again with the aggregated data
    aovrm = AnovaRM(data=aggregated_data, depvar='Total_Score', subject='Patient_ID', within=['Time', 'groupID'])
    res = aovrm.fit()
    result = res.summary()
except Exception as e:
    result = f"An error occurred: {e}"

result


# ANOVA Results Interpretation

Time (Effect of time on knowledge scores)

F Value: 0.0172
Degrees of Freedom (Num DF): 1
Degrees of Freedom (Den DF): 127
Pr > F (p-value): 0.8957
Interpretation: The effect of time (from T1 to T2) on knowledge scores is not statistically significant (p > 0.05), suggesting that knowledge scores did not significantly change over time across all participants.
groupID (Effect of being in the intervention vs. control group)

F Value: 32.5714
Degrees of Freedom (Num DF): 1
Degrees of Freedom (Den DF): 127
Pr > F (p-value): < 0.0001
Interpretation: The effect of group membership (intervention vs. control) is statistically significant, indicating a significant difference in knowledge scores between the intervention and control groups across both time points. The high F value suggests a strong effect.
Time:groupID (Interaction effect between time and groupID)

F Value: 0.6185
Degrees of Freedom (Num DF): 1
Degrees of Freedom (Den DF): 127
Pr > F (p-value): 0.4331
Interpretation: The interaction between time and groupID is not statistically significant, suggesting that the change in knowledge scores from T1 to T2 does not differ significantly between the intervention and control groups.

How to Proceed

Given these results:

Highlight the Group Effect: The significant difference between groups underscores the impact of the intervention. This is a key finding, as it suggests that participants in one group (presumably the intervention group) had different outcomes compared to the control group.

Consider the Lack of Time Effect: The lack of significant change over time suggests that, on average, knowledge scores did not increase from T1 to T2. This could be due to various factors, including the possibility that the intervention did not have the desired effect, or that the effect was not captured between these two time points.

Interaction Effect: The absence of a significant interaction effect indicates that the effect of the intervention did not significantly vary from T1 to T2 compared to the control. This could imply that the intervention's impact was consistent across time points or that any changes were not captured by the measurements used.

Further Analysis and Reporting: These results provide valuable insights into the effectiveness of the intervention and its consistency over time. When reporting these findings, it's important to discuss the context and implications of the significant group effect, the reasons behind the lack of significant time effect, and why the interaction might not have been significant. Additionally, considering other variables (as previously discussed) might provide further insights.

Explore Additional Variables: Given the significant group effect, exploring how other variables (e.g., Gravida, Para, country_of_origin) might influence the scores could provide deeper insights and help tailor future interventions more effectively.

# Finally, Why do we go back and choose Non-Parametric tests

Non-Normal Distribution: The Shapiro-Wilk tests indicated that the knowledge scores at both T1 and T2 are not normally distributed. Therefore, parametric tests such as the t-test, which assume normality, may not be appropriate. Instead, non-parametric tests that do not rely on the normality assumption are more suitable.

Comparison of Independent Groups: The Mann-Whitney U test is a non-parametric test used to compare the distribution of a continuous variable between two independent groups. In this case, we want to compare the knowledge scores between the treatment and control groups at both T1 and T2. The Mann-Whitney U test allows us to assess whether there are significant differences in knowledge levels between the two groups.

Comparison of Paired Groups: The Wilcoxon signed-rank test is a non-parametric test used to compare the distribution of a continuous variable between two paired groups. We can use this test to assess whether there are significant changes in knowledge scores within each group from T1 to T2. By comparing the scores within each group over time, we can evaluate the effectiveness of the intervention (IGVZ) in improving knowledge levels among pregnant asylum seekers.

Robustness to Outliers: Both the Mann-Whitney U test and the Wilcoxon signed-rank test are robust to outliers, which is important when dealing with real-world data that may contain extreme values or skewness.

Assumption of Random Sampling: Both tests do not assume that the data is drawn from a normally distributed population, making them suitable for the study's dataset, which may not adhere to strict assumptions of random sampling.

By choosing these non-parametric tests, we ensure that our statistical analyses are appropriate for the nature of the data and provide reliable results even when the normality assumption is violated.

Mann Whitney U Test

In [ ]:
from scipy.stats import mannwhitneyu

# Separate the data into treatment and control groups
treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Define the columns for T1 and T2 scores
t1_columns = [col for col in data.columns if 'T1_KQ_' in col and 'Score' in col]
t2_columns = [col for col in data.columns if 'T2_KQ_' in col and 'Score' in col]

# Function to print results
def print_test_results(group, col, stat, p_value, treatment_median, control_median):
    print(f"Mann-Whitney U test for {col} ({group}):")
    print(f"   U Statistic: {stat:.2f}")
    print(f"   p-value: {p_value:.4f}")
    print(f"   Treatment Group Median: {treatment_median}")
    print(f"   Control Group Median: {control_median}\n")

# Perform Mann-Whitney U test for T1 scores and compare medians
for col in t1_columns:
    stat, p_value = mannwhitneyu(treatment_group[col], control_group[col])
    treatment_median = treatment_group[col].median()
    control_median = control_group[col].median()
    print_test_results("T1", col, stat, p_value, treatment_median, control_median)

# Perform Mann-Whitney U test for T2 scores and compare medians
for col in t2_columns:
    stat, p_value = mannwhitneyu(treatment_group[col], control_group[col])
    treatment_median = treatment_group[col].median()
    control_median = control_group[col].median()
    print_test_results("T2", col, stat, p_value, treatment_median, control_median)


# Key Findings:

Most Scores Show No Significant Difference: Many of the p-values are above the conventional threshold of 0.05, indicating no statistically significant difference in the rank ordering of scores between the treatment and control groups for those questions. This is reflected in the median scores being equal for most questions.

Significant Differences in Specific Questions and Overall Scores: There are notable exceptions where the Mann-Whitney U test indicates a statistically significant difference between the groups. This includes:

T1_KQ_7_Score: Lower median in the treatment group (0.0 vs. 1.0, p=0.0159), suggesting the control group had higher scores.
T1_KQ_11_Score: No median difference (0.0 for both groups), but p=0.0063 indicates a difference in rank ordering.
T1_KQ_13b_Score: Equal medians (1.0 for both), but p=0.0067 suggests rank differences.
T1_KQ_13i_Score: Equal medians (1.0 for both), but p=0.0183 suggests rank differences.
T2_KQ_2_Score: Higher median in the treatment group (1.0 vs. 0.0, p=0.0001), indicating the treatment group had higher scores.
T2_KQ_3_Score, T2_KQ_4_Score, T2_KQ_7_Score, T2_KQ_10_Score, T2_KQ_12_Score, T2_KQ_13d_Score, T2_KQ_13g_Score, and T2_KQ_Total_Score show significant differences favoring the treatment group in several instances, particularly notable in the total scores for T2 (12.0 vs. 10.0, p=0.0133).

# Interpretation:

Treatment Effectiveness: The significant results, especially in T2 where the treatment group often shows a higher median score, suggest the intervention might have positively influenced certain knowledge areas. This is particularly evident in the total score for T2, indicating a broader impact of the treatment on overall knowledge by the second measurement point.

Rank Order vs. Median Scores: While the Mann-Whitney U test focuses on rank differences, observing the median scores provides a straightforward comparison of central tendency between groups. In cases where the p-value is significant but medians are the same, it suggests that while the median might not differ, the overall ranking of scores does, indicating a difference in distribution beyond the central point.

Reporting and Further Analysis: When reporting these findings, it’s crucial to note both the statistical significance (or lack thereof) and the practical significance, as indicated by differences in median scores. Where significant differences are observed, discussing the potential implications for the intervention's impact on knowledge acquisition will be key. Further, exploring why certain areas showed improvement while others did not could provide insights into refining or focusing future interventions.

This analysis highlights the nuanced impact of your intervention, demonstrating areas of significant knowledge gain and offering a basis for further investigation into the intervention's effects and potential improvements.


Wilcoxon Signed-Rank Test

In [ ]:
from scipy.stats import wilcoxon

# Perform Wilcoxon signed-rank test for T1 vs. T2 scores within treatment group
for t1_col, t2_col in zip(t1_columns, t2_columns):
    stat_treatment, p_value_treatment = wilcoxon(data[data['groupID'] == 1][t1_col], data[data['groupID'] == 1][t2_col])
    print(f"Wilcoxon signed-rank test for {t1_col} vs. {t2_col} (Treatment Group):")
    print(f"   Statistic: {stat_treatment}")
    print(f"   p-value: {p_value_treatment}")

# Perform Wilcoxon signed-rank test for T1 vs. T2 scores within control group
for t1_col, t2_col in zip(t1_columns, t2_columns):
    stat_control, p_value_control = wilcoxon(data[data['groupID'] == 0][t1_col], data[data['groupID'] == 0][t2_col])
    print(f"Wilcoxon signed-rank test for {t1_col} vs. {t2_col} (Control Group):")
    print(f"   Statistic: {stat_control}")
    print(f"   p-value: {p_value_control}")


In [ ]:
import pandas as pd
from scipy.stats import mannwhitneyu, chi2_contingency

# Load the data
#data = pd.read_csv('/mnt/data/data.csv')

# Define the parameters to compare
parameters = ['Years_of_education', 'Country_of_origin', 'Gravida', 'Para', 'T1_Gestational_age']

# Initialize a DataFrame to store the comparison results
comparison_results = pd.DataFrame(columns=['Parameter', 'Group', 'Median', 'IQR', 'P-value'])

# Function to perform the Mann-Whitney U test and return median, IQR, and p-value
def compare_groups(parameter):
    group1 = data[data['groupID'] == 1][parameter]
    group0 = data[data['groupID'] == 0][parameter]
    
    # Calculate median and interquartile range (IQR)
    median1, iqr1 = group1.median(), group1.quantile(0.75) - group1.quantile(0.25)
    median0, iqr0 = group0.median(), group0.quantile(0.75) - group0.quantile(0.25)
    
    # Perform the Mann-Whitney U test
    u_stat, p_val = mannwhitneyu(group1, group0)
    
    return median1, iqr1, median0, iqr0, p_val

# Perform comparisons for each parameter
for param in parameters:
    median1, iqr1, median0, iqr0, p_val = compare_groups(param)
    
    # Append the results for treatment group
    comparison_results = comparison_results.append({
        'Parameter': param,
        'Group': 'Treatment',
        'Median': median1,
        'IQR': iqr1,
        'P-value': p_val
    }, ignore_index=True)
    
    # Append the results for control group
    comparison_results = comparison_results.append({
        'Parameter': param,
        'Group': 'Control',
        'Median': median0,
        'IQR': iqr0,
        'P-value': p_val
    }, ignore_index=True)

comparison_results


In [ ]:
import pandas as pd
from scipy.stats import mannwhitneyu, chi2_contingency

# Load the data
#data = pd.read_csv('your_data.csv')  # Make sure to use your actual file path here

# Define a function to calculate medians, IQRs, and perform Mann-Whitney U test for continuous variables
def analyze_continuous(data, group_col, continuous_col):
    group_data = data.groupby(group_col)[continuous_col]
    medians = group_data.median()
    iqr = group_data.quantile(0.75) - group_data.quantile(0.25)
    u_stat, p_value = mannwhitneyu(data[data[group_col] == 1][continuous_col],
                                   data[data[group_col] == 0][continuous_col])
    return medians, iqr, p_value

# Define a function to create crosstabs and perform Chi-squared test for categorical variables
def analyze_categorical(data, group_col, category_col):
    crosstab = pd.crosstab(data[group_col], data[category_col])
    chi2, p_value, _, _ = chi2_contingency(crosstab)
    return crosstab, p_value

# Define your groups and parameters
groups = {1: 'Treatment', 0: 'Control'}
parameters = {
    'Years_of_education': 'continuous',
    'Country_of_origin': 'categorical',
    'Gravida': 'continuous',
    'Para': 'continuous',
    'T1_Gestational_age': 'continuous'
}

# Store the results
results = {}

# Analyze each parameter
for parameter, p_type in parameters.items():
    if p_type == 'continuous':
        medians, iqr, p_value = analyze_continuous(data, 'groupID', parameter)
        results[parameter] = {
            'Medians': medians,
            'IQRs': iqr,
            'P-value': p_value
        }
    elif p_type == 'categorical':
        crosstab, p_value = analyze_categorical(data, 'groupID', parameter)
        results[parameter] = {
            'Crosstab': crosstab,
            'P-value': p_value
        }

# Now, `results` contains all the analysis information
# You can print it or convert it into a DataFrame for a report
# For instance, to print the median and IQR for 'Years_of_education':
print('Years_of_education Medians:', results['Years_of_education']['Medians'])
print('Years_of_education IQRs:', results['Years_of_education']['IQRs'])
# To print the crosstab for 'Country_of_origin':
print('Country_of_origin Crosstab:\n', results['Country_of_origin']['Crosstab'])


In [ ]:
import pandas as pd
from scipy.stats import mannwhitneyu

# Load the data
#data = pd.read_csv('/mnt/data/data.csv')

# Define the parameters to compare
parameters = ['Years_of_education', 'Country_of_origin', 'Gravida', 'Para', 'T1_Gestational_age']

# Initialize an empty list to store the comparison results
comparison_results = []

# Function to perform the Mann-Whitney U test and return median, IQR, and p-value
def compare_groups(parameter):
    group1 = data[data['groupID'] == 1][parameter]
    group0 = data[data['groupID'] == 0][parameter]
    
    # Calculate median and interquartile range (IQR)
    median1, iqr1 = group1.median(), group1.quantile(0.75) - group1.quantile(0.25)
    median0, iqr0 = group0.median(), group0.quantile(0.75) - group0.quantile(0.25)
    
    # Perform the Mann-Whitney U test
    u_stat, p_val = mannwhitneyu(group1, group0)
    
    return median1, iqr1, median0, iqr0, p_val

# Perform comparisons for each parameter
for param in parameters:
    median1, iqr1, median0, iqr0, p_val = compare_groups(param)
    
    # Add the results for treatment group
    comparison_results.append({
        'Parameter': param,
        'Group': 'Treatment',
        'Median': median1,
        'IQR': iqr1,
        'P-value': p_val
    })
    
    # Add the results for control group
    comparison_results.append({
        'Parameter': param,
        'Group': 'Control',
        'Median': median0,
        'IQR': iqr0,
        'P-value': p_val
    })

# Convert the list of results to a DataFrame
comparison_results_df = pd.DataFrame(comparison_results)

comparison_results_df


In [ ]:
education = data.groupby(['groupID', 'Years_of_education']).agg('count').reset_index()
education




In [ ]:
# Define your parameters: Specify which are continuous and which are categorical
continuous_parameters = ['Years_of_education', 'T1_Gestational_age']
categorical_parameters = ['Country_of_origin', 'Gravida', 'Para']

# Assuming 'groupID' column with 1 for treatment and 0 for control group
treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Initialize a DataFrame to store results
results = []

# Function to perform Mann-Whitney U Test
def mann_whitney_u_test(group1, group2):
    stat, p = mannwhitneyu(group1, group2, alternative='two-sided')
    return p

# Analyze continuous parameters
for param in continuous_parameters:
    # Auto-binning continuous variables into quartiles
    data[param + '_bin'] = pd.qcut(data[param], q=4, duplicates='drop')
    
    # Iterate over each bin and calculate p-values
    for bin in data[param + '_bin'].unique():
        group1 = treatment_group[treatment_group[param + '_bin'] == bin][param]
        group2 = control_group[control_group[param + '_bin'] == bin][param]
        p_value = mann_whitney_u_test(group1, group2)
        results.append({'Parameter': param, 'Category/Bin': bin, 'P-value': p_value})
    
# Convert results to DataFrame for easy viewing
results_df = pd.DataFrame(results)

results_df

In [ ]:
import pandas as pd
from scipy.stats import mannwhitneyu, chi2_contingency
import numpy as np

# Assuming the data has been loaded into a DataFrame named `data`
# And 'groupID' indicates treatment (1) and control (0) groups

# Define continuous and categorical parameters
continuous_parameters = ['Years_of_education', 'T1_Gestational_age']
categorical_parameters = ['Country_of_origin', 'Gravida', 'Para']

# Prepare storage for the results
results = []

# Process continuous variables
for param in continuous_parameters:
    # Extract data for each group
    treatment_data = treatment_group[param].dropna()
    control_data = control_group[param].dropna()

    # Mann-Whitney U test to compare distributions
    if not treatment_data.empty and not control_data.empty:
        p_value = mannwhitneyu(treatment_data, control_data).pvalue
    else:
        p_value = np.nan  # Not applicable if one of the groups has no data

    results.append({
        'Parameter': param,
        'Category/Bin': 'Overall',
        'P-value': p_value,
        'Treatment Median': treatment_data.median(),
        'Control Median': control_data.median()
    })

# Process categorical variables
for param in categorical_parameters:
    # Get unique categories across both groups
    categories = np.union1d(treatment_group[param].unique(), control_group[param].unique())

    for category in categories:
        # Count occurrences in each group
        treatment_count = (treatment_group[param] == category).sum()
        control_count = (control_group[param] == category).sum()

        # Chi-squared test expects a contingency table as input
        contingency_table = np.array([[treatment_count, len(treatment_group) - treatment_count],
                                      [control_count, len(control_group) - control_count]])

        # Perform Chi-squared test if both groups have the category
        if treatment_count > 0 and control_count > 0:
            p_value = chi2_contingency(contingency_table, correction=False)[1]
        else:
            p_value = np.nan  # Not applicable if the category does not appear in one of the groups

        results.append({
            'Parameter': param,
            'Category/Bin': category,
            'P-value': "{:.17f}".format(p_value),
            'Treatment Count': treatment_count,
            'Control Count': control_count
        })

# Convert results to a DataFrame
results_df = pd.DataFrame(results)

results_df


In [ ]:
para = results_df[results_df['Parameter'] == 'Para'].drop(columns=['Treatment Median', 'Control Median']).reset_index()

para

In [ ]:
def generate_demographics(p, c):
    
    # Recategorize 'Para' and split data
    if p == 'Para':
        data['parameter'] = data[p].apply(categorize_para)
    elif p == 'Gravida':
        data['parameter'] = data[p].apply(categorize_para)
    elif p == 'T1_Gestational_age':
        data['parameter'] = data[p].apply(categorize_para)
    elif p == 'Years_of_education':
        data['parameter'] = data[p].apply(categorize_para)
    
    treatment_group = data[data['groupID'] == 1]
    control_group = data[data['groupID'] == 0]

    # Calculate frequencies for each category within both groups
    treatment_freq = treatment_group['parameter'].value_counts().reset_index(name='case_group')
    control_freq = control_group['parameter'].value_counts().reset_index(name='control_group')

    # Correctly rename 'index' column to merge on 'Para_category'
    treatment_freq.rename(columns={'index': 'parameter'}, inplace=True)
    control_freq.rename(columns={'index': 'parameter'}, inplace=True)

    # Merge treatment and control frequencies
    merged_freq = pd.merge(treatment_freq, control_freq, on='parameter', how='outer').fillna(0)

    # Calculate the Chi-square test for the merged frequencies
    chi2, p, dof, expected = chi2_contingency(merged_freq[['case_group', 'control_group']])

    # Add p-value to the merged frequencies DataFrame
    merged_freq['p-value'] = "{:.17f}".format(p)

    merged_freq['category'] = c

    new_order = ['category', 'parameter', 'case_group', 'control_group', 'p-value']

    # Rearrange columns using direct indexing
    merged_freq = merged_freq[new_order]

    # Now, df has columns ordered as specified in `new_order`
    # Display the merged frequencies along with the p-value
    print("Chi-square test p-value for differences across 'parameter':", p)
    return merged_freq.reset_index()

# Define the categories for "Para"
def categorize_para(value):
    if value == 0:
        return 'Nulliparous'
    elif 1 <= value <= 3:
        return 'Low Multiparous'
    elif value > 3:
        return 'Grand Multipara'
    else:
        return 'Unknown'
    
# Recategorize 'Gravida' and split data
def categorize_gravida(gravida):
    if gravida == 0:
        return 'Nulligravida'
    elif gravida == 1:
        return 'Primigravida'
    elif gravida == 2:
        return 'Secundigravida'
    elif 3 <= gravida <= 5:
        return 'Multigravida'
    else:
        return 'Unknown'

def categorize_gestational_age(gestational_age):
    if gestational_age < 28:
        return 'Extremely Preterm'
    elif 28 <= gestational_age < 32:
        return 'Very Preterm'
    elif 32 <= gestational_age < 37:
        return 'Moderate to Late Preterm'
    elif 37 <= gestational_age < 39:
        return 'Early Term'
    elif 39 <= gestational_age < 41:
        return 'Full Term'
    elif 41 <= gestational_age < 42:
        return 'Late Term'
    else:
        return 'Unknown'

def categorize_years_of_education(years):
    if years == 0:
        return 'No Formal Education'
    elif 1 <= years <= 6:
        return 'Primary Education'
    elif 7 <= years <= 12:
        return 'Secondary Education'
    elif 13 <= years <= 14:
        return 'Post-secondary Non-tertiary Education'
    elif 15 <= years <= 16:
        return 'Short-cycle Tertiary Education'
    elif 17 <= years <= 18:
        return 'Bachelor’s or Equivalent Level'
    elif 19 <= years <= 20:
        return 'Master’s or Equivalent Level'
    else:
        return 'Unknown'



In [ ]:
# Recategorize 'Para' and split data
data['parameter'] = data['Para'].apply(categorize_para)

treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Calculate frequencies for each category within both groups
treatment_freq = treatment_group['parameter'].value_counts().reset_index(name='case_group')
control_freq = control_group['parameter'].value_counts().reset_index(name='control_group')

# Correctly rename 'index' column to merge on 'Para_category'
treatment_freq.rename(columns={'index': 'parameter'}, inplace=True)
control_freq.rename(columns={'index': 'parameter'}, inplace=True)

# Merge treatment and control frequencies
merged_freq = pd.merge(treatment_freq, control_freq, on='parameter', how='outer').fillna(0)

# Calculate the Chi-square test for the merged frequencies
chi2, p, dof, expected = chi2_contingency(merged_freq[['case_group', 'control_group']])

# Add p-value to the merged frequencies DataFrame
merged_freq['p-value'] = "{:.3f}".format(p)

merged_freq['category'] = 'Para'

new_order = ['category', 'parameter', 'case_group', 'control_group', 'p-value']

# Rearrange columns using direct indexing
merged_freq = merged_freq[new_order]

# Now, df has columns ordered as specified in `new_order`
# Display the merged frequencies along with the p-value
print("Chi-square test p-value for differences across 'parameter':", p)
merged_freq.to_csv('/Users/nagesh/Library/CloudStorage/OneDrive-NextGenMetricsB.V/Personal/Anisha/M3/results.csv', mode='a', header=True, index=False)
merged_freq


In [ ]:
# Recategorize 'Gravida' and split data
def categorize_gravida(gravida):
    if gravida == 0:
        return 'Nulligravida'
    elif gravida == 1:
        return 'Primigravida'
    elif gravida == 2:
        return 'Secundigravida'
    elif 3 <= gravida <= 5:
        return 'Multigravida'
    else:
        return 'Grand multigravida'

# Apply the categorization to the 'Gravida' column
data['parameter'] = data['Gravida'].apply(categorize_gravida)

treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Calculate frequencies for each category within both groups
treatment_freq = treatment_group['parameter'].value_counts().reset_index(name='case_group')
control_freq = control_group['parameter'].value_counts().reset_index(name='control_group')

# Correctly rename 'index' column to merge on 'Para_category'
treatment_freq.rename(columns={'index': 'parameter'}, inplace=True)
control_freq.rename(columns={'index': 'parameter'}, inplace=True)

# Merge treatment and control frequencies
merged_freq = pd.merge(treatment_freq, control_freq, on='parameter', how='outer').fillna(0)

# Calculate the Chi-square test for the merged frequencies
chi2, p, dof, expected = chi2_contingency(merged_freq[['case_group', 'control_group']])

# Add p-value to the merged frequencies DataFrame
merged_freq['p-value'] = "{:.3f}".format(p)

merged_freq['category'] = 'Gravida'

new_order = ['category', 'parameter', 'case_group', 'control_group', 'p-value']

# Rearrange columns using direct indexing
merged_freq = merged_freq[new_order]

# Display the merged frequencies along with the p-value
print("Chi-square test p-value for differences across 'Gravida_category':", p)
merged_freq.to_csv('/Users/nagesh/Library/CloudStorage/OneDrive-NextGenMetricsB.V/Personal/Anisha/M3/results.csv', mode='a', header=False, index=False)

merged_freq


In [ ]:
def categorize_gestational_age(gestational_age):
    if gestational_age < 28:
        return 'Extremely Preterm'
    elif 28 <= gestational_age < 32:
        return 'Very Preterm'
    elif 32 <= gestational_age < 37:
        return 'Moderate to Late Preterm'
    elif 37 <= gestational_age < 39:
        return 'Early Term'
    elif 39 <= gestational_age < 41:
        return 'Full Term'
    elif 41 <= gestational_age < 42:
        return 'Late Term'
    else:
        return 'Post Term'

# Apply the categorization to the 'T1_Gestational_age' column
data['parameter'] = data['T1_Gestational_age'].apply(categorize_gestational_age)

treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Calculate frequencies for each category within both groups
treatment_freq = treatment_group['parameter'].value_counts().reset_index(name='case_group')
control_freq = control_group['parameter'].value_counts().reset_index(name='control_group')

# Correctly rename 'index' column to merge on 'Para_category'
treatment_freq.rename(columns={'index': 'parameter'}, inplace=True)
control_freq.rename(columns={'index': 'parameter'}, inplace=True)

# Merge treatment and control frequencies
merged_freq = pd.merge(treatment_freq, control_freq, on='parameter', how='outer').fillna(0)

# Calculate the Chi-square test for the merged frequencies
chi2, p, dof, expected = chi2_contingency(merged_freq[['case_group', 'control_group']])

# Add p-value to the merged frequencies DataFrame
merged_freq['p-value'] = "{:.3f}".format(p)

merged_freq['category'] = 'Gestational Age'

new_order = ['category', 'parameter', 'case_group', 'control_group', 'p-value']

# Rearrange columns using direct indexing
merged_freq = merged_freq[new_order]


# Display the merged frequencies along with the p-value
print("Chi-square test p-value for differences across 'T1_Gestational_age_category':", p)
merged_freq.to_csv('/Users/nagesh/Library/CloudStorage/OneDrive-NextGenMetricsB.V/Personal/Anisha/M3/results.csv', mode='a', header=False, index=False)
merged_freq


In [ ]:
def categorize_years_of_education(years):
    if years == 0:
        return 'No Formal Education'
    elif 1 <= years <= 6:
        return 'Primary Education'
    elif 7 <= years <= 12:
        return 'Secondary Education'
    elif 13 <= years <= 14:
        return 'Post-secondary Non-tertiary Education'
    elif 15 <= years <= 16:
        return 'Short-cycle Tertiary Education'
    elif 17 <= years <= 18:
        return 'Bachelor’s or Equivalent Level'
    elif 19 <= years <= 20:
        return 'Master’s or Equivalent Level'
    else:
        return 'Doctoral or Equivalent Level'

# Apply the categorization to the 'Years_of_education' column
data['parameter'] = data['Years_of_education'].apply(categorize_years_of_education)

treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Calculate frequencies for each category within both groups
treatment_freq = treatment_group['parameter'].value_counts().reset_index(name='case_group')
control_freq = control_group['parameter'].value_counts().reset_index(name='control_group')

# Correctly rename 'index' column to merge on 'Para_category'
treatment_freq.rename(columns={'index': 'parameter'}, inplace=True)
control_freq.rename(columns={'index': 'parameter'}, inplace=True)

# Merge treatment and control frequencies
merged_freq = pd.merge(treatment_freq, control_freq, on='parameter', how='outer').fillna(0)

# Calculate the Chi-square test for the merged frequencies
chi2, p, dof, expected = chi2_contingency(merged_freq[['case_group', 'control_group']])

# Add p-value to the merged frequencies DataFrame
merged_freq['p-value'] = "{:.3f}".format(p)

merged_freq['category'] = 'Gestational Age'

new_order = ['category', 'parameter', 'case_group', 'control_group', 'p-value']

# Rearrange columns using direct indexing
merged_freq = merged_freq[new_order]

# Display the merged frequencies along with the p-value
print("Chi-square test p-value for differences across 'Years_of_education_category':", p)
merged_freq.to_csv('/Users/nagesh/Library/CloudStorage/OneDrive-NextGenMetricsB.V/Personal/Anisha/M3/results.csv', mode='a', header=False, index=False)

merged_freq


In [ ]:
treatment_group = data[data['groupID'] == 1]
control_group = data[data['groupID'] == 0]

# Calculate frequencies for each category within both groups
treatment_freq = treatment_group['Country_of_origin'].value_counts().reset_index(name='case_group')
control_freq = control_group['Country_of_origin'].value_counts().reset_index(name='control_group')

# Correctly rename 'index' column to merge on 'Para_category'
treatment_freq.rename(columns={'index': 'Country_of_origin'}, inplace=True)
control_freq.rename(columns={'index': 'Country_of_origin'}, inplace=True)

# Merge treatment and control frequencies
merged_freq = pd.merge(treatment_freq, control_freq, on='Country_of_origin', how='outer').fillna(0)

# Calculate the Chi-square test for the merged frequencies
chi2, p, dof, expected = chi2_contingency(merged_freq[['case_group', 'control_group']])

# Add p-value to the merged frequencies DataFrame
merged_freq['p-value'] = "{:.3f}".format(p)

merged_freq['category'] = 'Country'

new_order = ['category', 'Country_of_origin', 'case_group', 'control_group', 'p-value']

# Rearrange columns using direct indexing
merged_freq = merged_freq[new_order]

# Display the merged frequencies along with the p-value
print("Chi-square test p-value for differences across 'Country_of_origin':", p)
merged_freq.to_csv('/Users/nagesh/Library/CloudStorage/OneDrive-NextGenMetricsB.V/Personal/Anisha/M3/results.csv', mode='a', header=False, index=False)

merged_freq


Correct Answer analysis

In [3]:

# Identify columns related to T1 scores
t1_score_columns = [col for col in data.columns if ('T1_' in col and '_Score' in col) or ('groupID' in col)]
t2_score_columns = [col for col in data.columns if ('T2_' in col and '_Score' in col) or ('groupID' in col)]

print(t1_score_columns)
print(t2_score_columns)

t1_correct = data.loc[:,t1_score_columns]
t1_correct['moment'] = 'T1'

# Remove prefix "T1_" and "T2_" and suffix "_Score" from column names
t1_correct.columns = t1_correct.columns.str.replace('T1_', '').str.replace('_Score', '').str.replace('KQ_', 'Question ')

t2_correct = data.loc[:,t2_score_columns]
t2_correct['moment'] = 'T2'

# Remove prefix "T1_" and "T2_" and suffix "_Score" from column names
t2_correct.columns = t2_correct.columns.str.replace('T2_', '').str.replace('_Score', '').str.replace('KQ_', 'Question ')

# Concatenate t1_correct and t2_correct DataFrames
correct = pd.concat([t1_correct, t2_correct])

correct

correct = correct.groupby(['groupID', 'moment']).agg('sum').reset_index()

correct.info()



['groupID', 'T1_KQ_1_Score', 'T1_KQ_2_Score', 'T1_KQ_3_Score', 'T1_KQ_4_Score', 'T1_KQ_5_Score', 'T1_KQ_6_Score', 'T1_KQ_7_Score', 'T1_KQ_8_Score', 'T1_KQ_9_Score', 'T1_KQ_10_Score', 'T1_KQ_11_Score', 'T1_KQ_12_Score', 'T1_KQ_13a_Score', 'T1_KQ_13b_Score', 'T1_KQ_13c_Score', 'T1_KQ_13d_Score', 'T1_KQ_13e_Score', 'T1_KQ_13f_Score', 'T1_KQ_13g_Score', 'T1_KQ_13h_Score', 'T1_KQ_13i_Score', 'T1_KQ_Total_Score']
['groupID', 'T2_KQ_1_Score', 'T2_KQ_2_Score', 'T2_KQ_3_Score', 'T2_KQ_4_Score', 'T2_KQ_5_Score', 'T2_KQ_6_Score', 'T2_KQ_7_Score', 'T2_KQ_8_Score', 'T2_KQ_9_Score', 'T2_KQ_10_Score', 'T2_KQ_11_Score', 'T2_KQ_12_Score', 'T2_KQ_13a_Score', 'T2_KQ_13b_Score', 'T2_KQ_13c_Score', 'T2_KQ_13d_Score', 'T2_KQ_13e_Score', 'T2_KQ_13f_Score', 'T2_KQ_13g_Score', 'T2_KQ_13h_Score', 'T2_KQ_13i_Score', 'T2_KQ_Total_Score']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          ------------

In [12]:
import pandas as pd

# Melt the 'correct' DataFrame
melted_correct = correct.melt(id_vars=['groupID', 'moment'], var_name='Question', value_name='Score')

# Pivot the melted DataFrame to get the desired format
result = melted_correct.pivot_table(index=['Question'], columns=['groupID', 'moment'], values='Score')

# Reset the index
result.reset_index(inplace=True)

# Now 'result' contains the desired DataFrame with Question as rows and GroupID and Moment as separate columns
result

groupID        Question      0             1        
moment                      T1     T2     T1      T2
0            Question 1   16.0   12.0   33.0    26.0
1           Question 10   11.0   11.0   30.0    59.0
2           Question 11   10.0    8.0    8.0     9.0
3           Question 12    8.0   15.0   17.0    19.0
4          Question 13a   27.0   23.0   60.0    68.0
5          Question 13b   35.0   22.0   68.0    69.0
6          Question 13c   16.0   13.0   34.0    44.0
7          Question 13d   17.0   13.0   41.0    54.0
8          Question 13e   16.0   13.0   23.0    38.0
9          Question 13f   26.0   20.0   52.0    67.0
10         Question 13g   24.0   15.0   55.0    70.0
11         Question 13h   27.0   21.0   60.0    72.0
12         Question 13i   31.0   22.0   58.0    70.0
13           Question 2    9.0    9.0   19.0    59.0
14           Question 3   25.0   17.0   56.0    66.0
15           Question 4   23.0   19.0   63.0    71.0
16           Question 5    3.0    4.0    9.0    18.0
17           Question 6   22.0   11.0   52.0    28.0
18           Question 7   23.0   14.0   36.0    63.0
19           Question 8    8.0    2.0   19.0    15.0
20           Question 9    9.0   10.0   34.0    39.0
21       Question Total  364.0  294.0  973.0  1024.0

In [13]:
result.info()
result

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   (Question, )  22 non-null     object 
 1   (0, T1)       22 non-null     float64
 2   (0, T2)       22 non-null     float64
 3   (1, T1)       22 non-null     float64
 4   (1, T2)       22 non-null     float64
dtypes: float64(4), object(1)
memory usage: 1012.0+ bytes


groupID        Question      0             1        
moment                      T1     T2     T1      T2
0            Question 1   16.0   12.0   33.0    26.0
1           Question 10   11.0   11.0   30.0    59.0
2           Question 11   10.0    8.0    8.0     9.0
3           Question 12    8.0   15.0   17.0    19.0
4          Question 13a   27.0   23.0   60.0    68.0
5          Question 13b   35.0   22.0   68.0    69.0
6          Question 13c   16.0   13.0   34.0    44.0
7          Question 13d   17.0   13.0   41.0    54.0
8          Question 13e   16.0   13.0   23.0    38.0
9          Question 13f   26.0   20.0   52.0    67.0
10         Question 13g   24.0   15.0   55.0    70.0
11         Question 13h   27.0   21.0   60.0    72.0
12         Question 13i   31.0   22.0   58.0    70.0
13           Question 2    9.0    9.0   19.0    59.0
14           Question 3   25.0   17.0   56.0    66.0
15           Question 4   23.0   19.0   63.0    71.0
16           Question 5    3.0    4.0    9.0    18.0
17           Question 6   22.0   11.0   52.0    28.0
18           Question 7   23.0   14.0   36.0    63.0
19           Question 8    8.0    2.0   19.0    15.0
20           Question 9    9.0   10.0   34.0    39.0
21       Question Total  364.0  294.0  973.0  1024.0

In [17]:
from scipy.stats import wilcoxon, mannwhitneyu

# Assuming `df` is your DataFrame and it's structured as described earlier
# Renaming columns for ease of access
result.columns = ['Question', 'Control_T1', 'Control_T2', 'Treatment_T1', 'Treatment_T2']
result.reset_index()

# Within-group comparison: Treatment group, T1 vs. T2 (to assess intervention effect)
stat, p_within_treatment = wilcoxon(result['Treatment_T1'], result['Treatment_T2'])

# Within-group comparison: Control group, T1 vs. T2 (to assess intervention effect)
stat, p_within_control = wilcoxon(result['Control_T1'], result['Control_T2'])

# Between-group comparison at T1 (baseline)
stat, p_between_t1 = mannwhitneyu(result['Control_T1'], result['Treatment_T1'])

# Between-group comparison at T2 (post-intervention)
stat, p_between_t2 = mannwhitneyu(result['Control_T2'], result['Treatment_T2'])

# Output the calculated p-values
print(f"Within-group p-value, Treatment (T1 vs. T2): {p_within_treatment}")
print(f"Within-group p-value, Control (T1 vs. T2): {p_within_control}")
print(f"Between-group p-value at T1: {p_between_t1}")
print(f"Between-group p-value at T2: {p_between_t2}")


Within-group p-value, Treatment (T1 vs. T2): 0.0006937980651855469
Within-group p-value, Control (T1 vs. T2): 0.0008734281806367522
Between-group p-value at T1: 0.0009256239560594297
Between-group p-value at T2: 1.6373263643966353e-05


/opt/homebrew/lib/python3.11/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: Exact p-value calculation does not work if there are zeros. Switching to normal approximation.
  res = hypotest_fun_out(*samples, **kwds)


In [18]:
result

Question  Control_T1  Control_T2  Treatment_T1  Treatment_T2
0       Question 1        16.0        12.0          33.0          26.0
1      Question 10        11.0        11.0          30.0          59.0
2      Question 11        10.0         8.0           8.0           9.0
3      Question 12         8.0        15.0          17.0          19.0
4     Question 13a        27.0        23.0          60.0          68.0
5     Question 13b        35.0        22.0          68.0          69.0
6     Question 13c        16.0        13.0          34.0          44.0
7     Question 13d        17.0        13.0          41.0          54.0
8     Question 13e        16.0        13.0          23.0          38.0
9     Question 13f        26.0        20.0          52.0          67.0
10    Question 13g        24.0        15.0          55.0          70.0
11    Question 13h        27.0        21.0          60.0          72.0
12    Question 13i        31.0        22.0          58.0          70.0
13      Question 2         9.0         9.0          19.0          59.0
14      Question 3        25.0        17.0          56.0          66.0
15      Question 4        23.0        19.0          63.0          71.0
16      Question 5         3.0         4.0           9.0          18.0
17      Question 6        22.0        11.0          52.0          28.0
18      Question 7        23.0        14.0          36.0          63.0
19      Question 8         8.0         2.0          19.0          15.0
20      Question 9         9.0        10.0          34.0          39.0
21  Question Total       364.0       294.0         973.0        1024.0

In [16]:
result.to_csv('/Users/nagesh/Library/CloudStorage/OneDrive-NextGenMetricsB.V/Personal/Anisha/M3/results.csv')